### PR TITLE
Improve documentation and error messages for safe initialization

### PR DESCRIPTION
**Pull request by liufengyun**
Improve documentation and error messages for safe initialization

### CODE REVIEW

**Comment by michelou** [comment]
@liufengyun Here are 3 changes to Markdown file [`safe-initialization.md`](https://github.com/lampepfl/dotty/pull/15659/commits/1aecc3f8ce43cf3e61b465199f491babed9a6830#diff-bee0bf17a51ba3930d7c8716dea79ff2acdaea5c91d2dbf72c75e526afdcfb7d)
| Old | New |
|:----|:----|
| The <ins>three</ins> principles above | The four principles above |
| Another exception <ins>too</ins> this rule  |  Another exception to this rule |
| essentially <ins>invovles</ins> close | essentially involves close |

File: docs/_docs/reference/other-new-features/safe-initialization.md [hunk 1]
```diff
@@ -159,41 +162,138 @@ Monotonicity is based on a well-known technique called _heap monotonic
...
+   sure the non-hot values are used properly.
+
+   In the above, `Vp` is the widened value of `p` --- the widening happens if `p`
+   is a warm value `Warm[D] { outer = V, args }` and we widen it to
+   `Warm[D] { outer = Cold, args }`.
+
+   The variable `Vargs` represents values of `args` with non-hot values widened
+   to `Cold`.
+
+   The widening is motivated to finitize the abstract domain and ensure
```
Comment (olhotak) [inline docs/_docs/reference/other-new-features/safe-initialization.md:280]:
```suggestion
   The motivation for the widening is to finitize the abstract domain and ensure
```

File: docs/_docs/reference/other-new-features/safe-initialization.md [hunk 2]
```diff
@@ -159,41 +162,138 @@ Monotonicity is based on a well-known technique called _heap monotonic
...
+   terimination of the initialization check.
```
Comment (olhotak) [inline docs/_docs/reference/other-new-features/safe-initialization.md:281]:
```suggestion
   termination of the initialization check.
```

File: docs/_docs/reference/other-new-features/safe-initialization.md [hunk 3]
```diff
@@ -159,41 +162,138 @@ Monotonicity is based on a well-known technique called _heap monotonic
...
+
+9. The scrutinee in pattern match, the values in return and throw statements must be _effectively hot_.
```
Comment (olhotak) [inline docs/_docs/reference/other-new-features/safe-initialization.md:283]:
```suggestion
9. The scrutinee in a pattern match and the values in return and throw statements must be _effectively hot_.
```

File: docs/_docs/reference/other-new-features/safe-initialization.md [hunk 4]
```diff
@@ -159,41 +162,138 @@ Monotonicity is based on a well-known technique called _heap monotonic
...
+
+A value `v` is _effectively hot_ if any of the following is true:
+
+- `v` is `Hot`.
+- `v` is `ThisRef` and all fields of the underlying object are initialized.
+- `v` is `Warm[C] { outer = V, ctor, args = Vs }` and
+  1. `C` does not contain inner classes;
+  2. Calling any method on `v` encounters no errors and the method return value is _effectively hot_;
```
Comment (olhotak) [inline docs/_docs/reference/other-new-features/safe-initialization.md:291]:
```suggestion
  2. Calling any method on `v` encounters no initialization errors and the method return value is _effectively hot_;
```

Comment (olhotak) [inline docs/_docs/reference/other-new-features/safe-initialization.md:291]:
```suggestion
  2. Calling any method on `v` encounters no errors and the method return value is _effectively hot_; and
```

File: docs/_docs/reference/other-new-features/safe-initialization.md [hunk 5]
```diff
@@ -159,41 +162,138 @@ Monotonicity is based on a well-known technique called _heap monotonic
...
+
+   The variable `Vargs` represents values of `args` with non-hot values widened
+   to `Cold`.
+
+   The widening is motivated to finitize the abstract domain and ensure
+   terimination of the initialization check.
+
+9. The scrutinee in pattern match, the values in return and throw statements must be _effectively hot_.
+
+A value `v` is _effectively hot_ if any of the following is true:
```
Comment (olhotak) [inline docs/_docs/reference/other-new-features/safe-initialization.md:285]: Move this definition before "Rules".

File: docs/_docs/reference/other-new-features/safe-initialization.md [hunk 6]
```diff
@@ -159,41 +162,138 @@ Monotonicity is based on a well-known technique called _heap monotonic
...
+
+- `v` is `Hot`.
+- `v` is `ThisRef` and all fields of the underlying object are initialized.
```
Comment (olhotak) [inline docs/_docs/reference/other-new-features/safe-initialization.md:288]:
```suggestion
- `v` is `ThisRef` and all fields of the underlying object are assigned.
```

File: docs/_docs/reference/other-new-features/safe-initialization.md [hunk 7]
```diff
@@ -159,41 +162,138 @@ Monotonicity is based on a well-known technique called _heap monotonic
...
+- `v` is `Warm[C] { outer = V, ctor, args = Vs }` and
+  1. `C` does not contain inner classes;
+  2. Calling any method on `v` encounters no errors and the method return value is _effectively hot_;
+  3. Each field of `v` is _effectively hot_.
+- `v` is `Fun(e, V, C)` and calling the function encounters no errors and the
+  function return value is _effectively hot_.
+- `ThisRef` is _effectively hot_.
+
```
Comment (olhotak) [inline docs/_docs/reference/other-new-features/safe-initialization.md:296]:
```suggestion
In the `ThisRef` case, local reasoning ensures that all fields assigned implies that all fields themselves contain effectively hot values.

In the `Warm` case, if the value `v` leaks outside the analysis scope, the unanalyzed code could instantiate inner classes of `v`, call methods on `v`, or read fields of `v`. The conditions of the `Warm` case ensure that these actions do not lead to accessing a cold object.

In the `Fun` case, ...
```

**Review by liufengyun — COMMENTED** [review]
Apply suggested changes

**Review by liufengyun — COMMENTED** [review]
Apply suggested changes

File: docs/_docs/reference/other-new-features/safe-initialization.md (hunk 7 above)
Comment (liufengyun) [reply docs/_docs/reference/other-new-features/safe-initialization.md:302]: Maybe we can add the justification later --- they are not immediately understandable to readers and can itself be an obstacle in reading the docs.

**Comment by liufengyun** [comment]
Thanks @michelou , I fixed two of them. Regarding

> The three principles above

The usage of "three" is intended here and it's not a typo.

**Pull request merged**